Add support for audit.v3 logging

nmiyake opened the pull request:
Does the following:
* Adds the audit3log package with functions to create and log audit.3 logs and support for dual-writing audit.2 logs
* Adds dual-writing capability to audit.2 logs so that logging audit.2 logs will dual-write an equivalent audit.3 log entry

## Before this PR
<!-- What's wrong with the current state of the world and why change it now? -->

## After this PR
<!-- User-facing outcomes this PR delivers go below -->
==COMMIT_MSG==
Adds support for audit.v3 logging
==COMMIT_MSG==

## Possible downsides?
<!-- Please describe any way users could be negatively affected by this PR. -->

nmiyake committed:
Add support for audit.v3 logging

Does the following:
* Adds the audit3log package with functions to create and log
  audit.3 logs and support for dual-writing audit.2 logs
* Adds dual-writing capability to audit.2 logs so that logging
  audit.2 logs will dual-write an equivalent audit.3 log entry

bmoylan commented on wlog/auditlog/internal/auditloginternal/audit3_logger.go:
```diff
@@ -0,0 +1,67 @@
+	Reason string `json:"reason"`
+}
+
+type Audit3ContextualizedUser struct {
```
Do we want `omitempty` on the json struct tags to avoid a bunch of empty fields?

nmiyake replied:
ACK -- makes sense for the optional fields. Done.

bmoylan reviewed (dismissed):
This looks good! Thanks for putting it together. Approving because my question is nonblocking and if you decide to make the proposed change I think it can merge as is.

IlanaRadinsky commented on wlog/auditlog/audit3log/params.go:
```diff
@@ -0,0 +1,196 @@
+	StackKey          = auditloginternal.Audit3StackKey
+	ServiceKey        = auditloginternal.Audit3ServiceKey
+	EnvironmentKey    = auditloginternal.Audit3EnvironmentKey
+	ProducerTypeKey   = auditloginternal.Audit3ProducerTypeKey
```
This can probably always be `SERVER`

nmiyake replied:
ACK -- even so, this is defining the data structure, so as long as it's included in the output this still needs to be here

IlanaRadinsky commented on wlog/auditlog/audit3log/params.go:
```diff
@@ -0,0 +1,196 @@
+	EnvironmentKey    = auditloginternal.Audit3EnvironmentKey
+	ProducerTypeKey   = auditloginternal.Audit3ProducerTypeKey
+	OrganizationsKey  = auditloginternal.Audit3OrganizationsKey
+	EventIDKey        = auditloginternal.Audit3EventIDKey
```
Also `logEntryId`?

nmiyake replied:
OK thanks -- I see that documented at https://www.palantir.com/docs/foundry/security/audit-logs-overview#audit3-logs-not-yet-available, but it doesn't seem to be part of the Conjure API or sls-spec, which is what I used to do the implementation here.

IlanaRadinsky replied:
`logEntryId` is in the latest conjure API, but it's not in sls-spec - I'll track getting it added there.

IlanaRadinsky replied:
`logEntryId` is always set in foundry-audit V3, so we should probably set it here too, even though we don't currently support emitting multiple audit logs for a single `eventId` at the moment.

https://github.palantir.build/foundry/foundry-audit/blob/baf25b651addacd7fb83ccd43c1ad7a36cb27d0e/audit-core-api/src/main/java/com/palantir/foundry/audit/AuditV3Logger.java#L161

nmiyake replied:
Thanks! After looking at usage here a bit more, these constants are only used by services that parse the logs directly (like `planer`), so reducing the exported constants to the minimum required ones for now:

* `Type`
* `Name`
* `Result`
* `LogEntryId`

Can always add/export more later if there is need/demand.

IlanaRadinsky commented on wlog/auditlog/audit3log/params.go:
```diff
@@ -0,0 +1,196 @@
+	ServiceKey        = auditloginternal.Audit3ServiceKey
+	EnvironmentKey    = auditloginternal.Audit3EnvironmentKey
+	ProducerTypeKey   = auditloginternal.Audit3ProducerTypeKey
+	OrganizationsKey  = auditloginternal.Audit3OrganizationsKey
```
We actually don't populate this field in v3 logs, so you can probably leave it out.

nmiyake replied:
This comes from the `sls-logging-api` YAML. Is this field officially deprecated/unused? If so, can we update the documentation in the Conjure definition to reflect this?

IlanaRadinsky replied:
Will track that!

IlanaRadinsky commented on wlog/auditlog/audit3log/context.go:
```diff
@@ -0,0 +1,66 @@
+func FromContext(ctx context.Context) Logger {
+	logger := loggerFromContext(ctx)
+	var params []Param
+	if uid := wloginternal.IDFromContext(ctx, wloginternal.UIDKey); uid != nil {
+		params = append(params, UID(*uid))
+	}
+	if sid := wloginternal.IDFromContext(ctx, wloginternal.SIDKey); sid != nil {
+		params = append(params, SID(*sid))
+	}
+	if tokenID := wloginternal.IDFromContext(ctx, wloginternal.TokenIDKey); tokenID != nil {
+		params = append(params, TokenID(*tokenID))
+	}
+	if orgID := wloginternal.IDFromContext(ctx, wloginternal.OrgIDKey); orgID != nil {
+		params = append(params, OrgID(*orgID))
+	}
```
Just curious where the logic that pulls these values out of the token and adds them to the `Context` lives?

nmiyake replied:
That would be in `witchcraft-go-logging` or another middleware layer.

nmiyake replied:
PR is up -- you can see that the logic at https://github.com/palantir/witchcraft-go-server/pull/942/files#diff-53c6f527063e7a4d77c40ccbad59b10b026d99ad006df6d4af0e4d844c657ee2R215 sets the audit logger so that it uses values set on the context extracted from the token

IlanaRadinsky commented on wlog/auditlog/internal/auditloginternal/logger.go:
```diff
@@ -28,8 +31,56 @@ const (
+	// dual-log if audit3Logger specified
+	if l.audit3Logger != nil {
+		audit3Params := make([]Audit3Param, len(params))
+		for i, param := range params {
+			audit3Params[i] = Audit3Param{
+				Param: param.Param,
+			}
+		}
```
Does this only convert an `Audit2Param` to an `Audit3Param` if they have the same underlying type of `AuditParam`? But if you use the `audit.2` logger, the `audit.3` log that is dual-written will only contain fields that are in the `audit.2` log? We'd probably need to make sure that all the fields required on `AuditLogV3` are present. You mentioned in the existing state, clients provide the category as a `RequestParam` - when converting to `audit.3`, can we move any categories from the `RequestParam` field to the `categories` field? Also, we can consider moving values from `Audit2RequestParams` and `Audit2ResultParams` to `Audit3RequestFields` and `Audit3ResultFields`.

Also, is there a way to add fields that foundry-audit will pull from request headers - `origins`, `sourceOrigins`, `userAgent`? Not sure how these are currently populated in `audit.2` logs. In foundry-audit, they're added to the requestParams

Do existing clients of the `audit.2` logger provide categories?

In `audit.2`, clients provide `requestParams` and `resultParams` which might contain `category` or `categories` fields. If we wanted to dual-write to `audit.3` type, we'd need to pull the

nmiyake replied:
A few different things here...

The current design in this PR (and the previous one) is basically that we define a single `AuditParam` interface (https://github.com/palantir/witchcraft-go-logging/blob/7a34904962d261facd7b5318ac7c92bddd81a12e/wlog/auditlog/internal/auditloginternal/params.go#L21-L24) -- parameters that implement that interface must define how the parameter is logged for `audit.2` and `audit.3` (which can also be a noop).

For example, the `Audit2RequestParams` parameter (https://github.com/palantir/witchcraft-go-logging/blob/7a34904962d261facd7b5318ac7c92bddd81a12e/wlog/auditlog/internal/auditloginternal/audit2_params.go#L110-L121) is defined in such a manner that it logs to the `requestParams` field for `audit.2`, but to the `requestFields` field for `audit.3`.

The request header fields makes sense, although that will likely be more of a middleware/server/HTTP client concern. The logic would basically make it so that, at that middleware level, we extract these relevant fields from requests and then set them on any audit loggers that are present in the context.

> Do existing clients of the audit.2 logger provide categories?

I don't believe that we have any Go code that produces audit logs with any category information. I see some documentation on v2 categories at https://www.palantir.com/docs/foundry/security/audit-log-categories, but no mention for V2 in the spec or Conjure definitions.

nmiyake replied:
> Also, is there a way to add fields that foundry-audit will pull from request headers - origins, sourceOrigins, userAgent? Not sure how these are currently populated in audit.2 logs. In foundry-audit, they're added to the requestParams

In the current design, it looks like our audit middleware library logs every key-value pair in the header (except for certain hard-coded keys, for which values are redacted) as a map under the `header` key

IlanaRadinsky replied:
> > Do existing clients of the audit.2 logger provide categories?
>
> I don't believe that we have any Go code that produces audit logs with any category information. I see some documentation on v2 categories at https://www.palantir.com/docs/foundry/security/audit-log-categories, but no mention for V2 in the spec or Conjure definitions.

In foundry-audit, it's possible to include a `category` object which will be logged as a request param

And so foundry-audit has logic to include the `category` fields in the corresponding fields for audit.3 logs

It does look like categories are used by some Go services so would be ideal to have parallel logic here.

IlanaRadinsky replied:
>> Also, is there a way to add fields that foundry-audit will pull from request headers - origins, sourceOrigins, userAgent? Not sure how these are currently populated in audit.2 logs. In foundry-audit, they're added to the requestParams
>
> In the current design, it looks like our audit middleware library logs every key-value pair in the header (except for certain hard-coded keys, for which values are redacted) as a map under the header key

Would it be possible to update the middleware to add the expected audit log fields based on the request context? We have such logic in `foundry-audit`.

nmiyake committed:
Add changelog

IlanaRadinsky commented on wlog/auditlog/audit3log/category.go:
```diff
@@ -0,0 +1,1647 @@
+type auditCategoryV2Visitor struct {
+}
+
+func (a *auditCategoryV2Visitor) VisitDataCreate(ctx context.Context, v v2.DataCreate) (Param, error) {
```
It's unfortunate that this all has to be done manually... For Typescript support, we added the logic to convert a category to request/result params in the foundry-audit repo: https://github.palantir.build/foundry/foundry-audit/tree/develop/audit-categories-api/audit-categories-api-typescript and then added tests within the repo to assert that the Typescript implementation produced the same results as the Java implementation: https://github.palantir.build/foundry/foundry-audit/blob/develop/audit-core-api/src/test/java/com/palantir/foundry/audit/categories/AuditCategoryConfigurationCoverageTest.java#L819-L911 . Would it be possible/useful to do the same for the Go implementation?

nmiyake replied:
I don't think there's really a way around doing this "manually" somewhere, is there? Isn't this the equivalent of https://github.palantir.build/foundry/foundry-audit/blob/develop/audit-categories-api/audit-categories-api-typescript/src/auditCategoryConfiguration.ts#L211 or https://github.palantir.build/foundry/foundry-audit/blob/c28c67f8d21bf558c27d0cdd9614ca5341e1d7ea/audit-categories-api/src/main/java/com/palantir/foundry/audit/categories/AuditCategoryConfiguration.java#L133?

Basically, for a given V2 category, we need to know the request vs. result parameters, and I don't think that's encoded anywhere in code (I was going off the table on one of the web sites), so we have to have to "manually" encode this information *somewhere*.

This could be restructured to be more like the Java code and to have a `AuditCategoryConfiguration` equivalent that encodes this information and then have the visitor code work off of that -- I guess that would centralize the definition and reduce the risk of human error in the individual functions (in matching `audit2` and `audit3` param code and ensuring that all fields that are declared in config are converted in visitors).

Also happy to take a look and adopt/run the same test cases.

IlanaRadinsky commented on wlog/auditlog/audit3log/params.go:
```diff
@@ -0,0 +1,201 @@
+	return convertInternalParamToExportedParam(auditloginternal.Audit3Entities(entities))
+}
+
+func Users(users []ContextualizedUser) Param {
+	var audit3ContextualizedUsers []auditloginternal.Audit3ContextualizedUser
+	if len(users) > 0 {
+		audit3ContextualizedUsers = make([]auditloginternal.Audit3ContextualizedUser, len(users))
+		for i, user := range users {
+			audit3ContextualizedUsers[i] = auditloginternal.Audit3ContextualizedUser(user)
+		}
+	}
+	return convertInternalParamToExportedParam(auditloginternal.Audit3Users(audit3ContextualizedUsers))
+}
```
We also expect that the `userId` extracted from the token should be added to the `Users` field at some point. https://github.palantir.build/foundry/foundry-audit/blob/develop/audit-core-api/src/main/java/com/palantir/foundry/audit/AuditV3Logger.java#L114-L117

nmiyake replied:
Thanks for flagging -- great catch, and didn't know about this/see this. Will verify that this gets added (either in this project or in the `witchcraft-go-server` code)

IlanaRadinsky commented on wlog/auditlog/audit3log/params.go:
```diff
@@ -0,0 +1,201 @@
+	return convertInternalParamToExportedParam(auditloginternal.Audit3Categories(categories))
+}
+
+func Entities(entities []any) Param {
+	return convertInternalParamToExportedParam(auditloginternal.Audit3Entities(entities))
+}
```
For entities, foundry-audit will extract rids from `Resource`s and only include the `rid`s in the audit log: https://github.palantir.build/foundry/foundry-audit/blob/develop/audit-core-api/src/main/java/com/palantir/foundry/audit/api/common/v2/EntityExtractor.java#L10 https://github.palantir.build/foundry/foundry-audit/blob/baf25b651addacd7fb83ccd43c1ad7a36cb27d0e/audit-core-api/src/main/java/com/palantir/foundry/audit/AuditLogEntryImpl.java#L182 Is it possible to do the same here

nmiyake replied:
Thanks, will take a look

IlanaRadinsky commented on wlog/wrappedlog/wrapped1log/wrapped1logtests/audit3log_tests.go:
```diff
@@ -0,0 +1,338 @@
+			RequestFields: map[string]any{"requestKey": "requestValue"},
+			ResultFields:  map[string]any{"resultKey": "resultValue"},
+			JSONMatcher: map[string]objmatcher.Matcher{
+				"type":          objmatcher.NewEqualsMatcher("wrapped.1"),
+				"entityName":    objmatcher.NewEqualsMatcher(entityName),
+				"entityVersion": objmatcher.NewEqualsMatcher(entityVersion),
```
foundry-audit doesn't currently emit audit logs as wrapped logs, but this should still be fine since vector will unwrap the logs. But maybe worth emitting them as unwrapped logs to be consistent?

What will the entity name and version be?

nmiyake replied:
I'm honestly not 100% sure -- this was mainly just matching existing structure for other wrapped logs. I think that some of our logging infrastructure might used this wrapped log stuff, but will double-check.

nmiyake reviewed (commented):
Thanks for the review/feedback! Will update based on this feedback and echo back when ready for another review.

nmiyake committed:
Update PR feedback

nmiyake committed:
Add ObjectListValue and ObjectListAppendValue to LogEntry

Update the LogEntry interface to add the "ObjectListValue" and
"ObjectListAppendValue" functions. These functions are analogous
to the existing "StringListValue" and "StringListAppendValue" functions
and makes it possible to support log entry types for append-only lists
for non-string types.

nmiyake committed:
Update tests

nmiyake committed:
Update tests and logic

nmiyake committed:
Update generation

nmiyake committed:
Update formatting

nmiyake committed:
Update generate

nmiyake commented:
@IlanaRadinsky thanks for the initial review! I've updated the PR to address the core points of feedback. Main changes:

* Reduced the amount of manual code written for categories. Instead, wrote a program that parses the contents of the Java `AuditCategoryConfiguration.java` file and generates the equivalent Go code for the categories defined in that file. This automates the process of generating categories and ensures that things like required vs. optional fields and classifications match the Java definitions.
* Added extractor logic for entities (RIDs) and UIDs that mirrors the Java logic. Like the Java logic, the default implementation parses through fields. There is also support for custom extraction per category type, but that must be defined manually. Matched the current Java behavior where some categories have custom user extractors, but all resource extractors are just the defaults.

In order to properly support entities and user IDs (which are non-strings but must be appendable), will need #422 to merge first (this PR includes that change to make it build and to unblock review).

nmiyake commented on wlog/auditlog/audit3log/categories.go:
```diff
@@ -0,0 +1,785 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Code generated by category_transform.go using source file version 6.101.0, DO NOT EDIT.
```
As noted here, this is a generated file, so every line does not need to be reviewed.

However, it would be helpful to do a spot check to ensure that the general logic that is generated makes sense/is correct.

nmiyake commented on wlog/auditlog/audit3log/audit3logtests/tests.go:
```diff
@@ -0,0 +1,757 @@
+			},
+		},
+		{
+			TestCaseName: "audit log entry sets values extracted from categories",
```
Verifies extraction

nmiyake commented on wlog/auditlog/audit3log/category.go:
```diff
@@ -0,0 +1,145 @@
+						ResultField(k, fieldVal.extractField(v)).getParam().Audit3ParamFn(entry)
+					}
+
+					// add all extracted entities
+					entityExtractorFn := c.entityExtractor
+					if entityExtractorFn == nil {
+						entityExtractorFn = defaultEntityExtractor[T]
+					}
+					if entities, _ := entityExtractorFn(v, c); len(entities) > 0 {
+						anyEntriesSlice := make([]any, len(entities))
+						for i, entity := range entities {
+							anyEntriesSlice[i] = entity
+						}
+						Entities(anyEntriesSlice).getParam().Audit3ParamFn(entry)
+					}
+
+					// add all extracted user IDs
+					userIDExtractorFn := c.uidExtractor
+					if userIDExtractorFn == nil {
+						userIDExtractorFn = defaultUIDExtractor[T]
+					}
+					if userIDs, _ := userIDExtractorFn(v, c); len(userIDs) > 0 {
+						var users []ContextualizedUser
+						for _, userID := range userIDs {
+							users = append(users, ContextualizedUser{
+								UID: string(userID),
+							})
+						}
+						Users(users).getParam().Audit3ParamFn(entry)
+					}
```
This is the logic that extracts entities and user IDs from category values

nmiyake commented on wlog/auditlog/audit3log/internal/categories/category.go:
```diff
@@ -0,0 +1,137 @@
+)
+
+// Converts the provided data to a slice of v2.Resource. Based on the Java "promote" method.
+func toResource(data any) ([]commonv2.Resource, error) {
```
This file and all functions are based very closely on implementation/logic in `EntityExtractor.java`

nmiyake commented on wlog/auditlog/audit3log/internal/categories/entityextractor.go:
```diff
@@ -0,0 +1,209 @@
+	commonv2 "github.com/palantir/witchcraft-go-logging/conjure/foundry/audit/api/common/v2"
+)
+
+func EntitiesFromResources(resource ...commonv2.Resource) ([]rid.ResourceIdentifier, error) {
```
Logic in this file also mostly analogous to `EntityExtractor.java`

nmiyake commented on wlog/auditlog/audit3log/internal/category_transform.go:
```diff
@@ -0,0 +1,224 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
```
This is the program that takes a `AuditCategoryConfiguration.java` file and generates the `categories.go` file with equivalent definitions

nmiyake commented on wlog/auditlog/audit3log/categories_uid_extractors.go:
```diff
@@ -0,0 +1,72 @@
+	}
+}
+
+func defaultUIDExtractor[T any](val T, cat *categoryInfo[T]) ([]logging.UserId, error) {
```
Default extractor logic based on Java logic

nmiyake commented on wlog/auditlog/audit3log/categories_uid_extractors.go:
```diff
@@ -0,0 +1,72 @@
+var _ uidExtractor[any] = defaultUIDExtractor[any]
+
+func init() {
+	categoryManagementTokens.uidExtractor = func(val v2.ManagementTokens, cat *categoryInfo[v2.ManagementTokens]) ([]logging.UserId, error) {
+		return categoriespkg.ExtractUserIDsFromTokens(val.ManagedTokens), nil
+	}
+	categoryPassThrough.uidExtractor = func(val v2.PassThrough, cat *categoryInfo[v2.PassThrough]) ([]logging.UserId, error) {
+		return nil, nil
+	}
+	categoryTokenAccess.uidExtractor = func(val v2.TokenAccess, cat *categoryInfo[v2.TokenAccess]) ([]logging.UserId, error) {
+		return categoriespkg.ExtractUserIDsFromTokens(val.AccessedTokens), nil
+	}
+	categoryTokenGeneration.uidExtractor = func(val v2.TokenGeneration, cat *categoryInfo[v2.TokenGeneration]) ([]logging.UserId, error) {
+		return categoriespkg.ExtractUserIDsFromTokens(val.GeneratedTokens), nil
+	}
+	categoryTokenRevoke.uidExtractor = func(val v2.TokenRevoke, cat *categoryInfo[v2.TokenRevoke]) ([]logging.UserId, error) {
+		return categoriespkg.ExtractUserIDsFromTokens(val.RevokedTokens), nil
+	}
```
These custom UID extractors are based on the custom ones set in the Java code.

These custom extractors do need to be set/updated manually. Will likely need to update the documentation/instructions for adding a new category type to include modifying this file.

nmiyake commented on wlog/auditlog/audit3log/categories_entity_extractors.go:
```diff
@@ -0,0 +1,58 @@
+
+var _ entityExtractor[any] = defaultEntityExtractor[any]
+
+func defaultEntityExtractor[T any](val T, cat *categoryInfo[T]) ([]rid.ResourceIdentifier, error) {
```
Default extractor logic based on Java logic

nmiyake commented on wlog/auditlog/audit3log/categories_entity_extractors.go:
```diff
@@ -0,0 +1,58 @@
+	commonv2 "github.com/palantir/witchcraft-go-logging/conjure/foundry/audit/api/common/v2"
+	categoriespkg "github.com/palantir/witchcraft-go-logging/wlog/auditlog/audit3log/internal/categories"
+)
+
```
Checked Java definition/implementation, and currently there are no custom entity extractors (all entity extractors use the default logic), although the visitor logic does allow for adding a custom extractor.

In Go, kept this logic structured in the same manner as UID extractor, so if a category that is added in the future were to need a custom entity extractor, it could be set up in an `init` function here in the same manner as for UID extractors.

IlanaRadinsky reviewed (dismissed):
LGTM aside from one comment!

nmiyake committed:
Update params

nmiyake committed:
Merge remote-tracking branch 'origin/develop' into nmiyake/add-audit-3-04-30-2025-submit-with-append

bmoylan reviewed (approved):
Sorry for the delay in reviewing, but lgtm!
